modules/replicator/src/main/java/org/apache/ignite/internal/replicator/ReplicaManager.java

```diff
@@ -217,20 +217,24 @@ private Replica startReplicaInternal(
             ReplicationGroupId replicaGrpId,
             ReplicaListener listener
     ) {
+        Replica newReplica = new Replica(replicaGrpId, listener);
+
         replicas.compute(replicaGrpId, (replicationGroupId, replicaFut) -> {
             if (replicaFut == null) {
-                replicaFut = CompletableFuture.completedFuture(new Replica(replicaGrpId, listener));
-
-                return replicaFut;
+                return CompletableFuture.completedFuture(newReplica);
             } else {
-                replicaFut.complete(new Replica(replicaGrpId, listener));
+                if (replicaFut.isDone() && !replicaFut.isCancelled() && !replicaFut.isCompletedExceptionally()) {
```

Comment:
When can this happen? Are you sure we can simply ignore the previous replica value?

Reply:
It can happen on rebalance, then we start a new replica while we still have an old replica. It looks like we can, because it doesn't have a state (also, note that in `stopReplica` we just remove the object from the map)

Reply:
But how can we have an old replica in the map? Why wasn't `stopReplica` called?

Reply:
Because the node was either in pending or in stable assignments

Reply:
Can you please describe this code path?

Reply:
Sure: if we are doing rebalancing and `STABLE_ASSIGNMENTS_*` key changed in metastorage, then we need to decide if replica must be stopped. We will stop it only if the node is not in stableAssignments and not in pendingAssignments. We do start replica on PENDING_ASSIGNMENTS_* and in `updateAssignmentInternal()`, so the replica may be not stopped.